Adds Preferences To Suppress Ghost Role Rolls

san7890 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Hey there,

Ever since November of 2021, I've wanted something where I could simply not get any ghost roles while adminned. Some people also do not want to get any ghost rolls whatsoever when they play, for it is their personal preference. This PR seeks to resolve both of these issues with two new preferences.

The first preference will show up to everyone, Suppress All Ghost Rolls. It will return on the main proc that pops up the window, does the sound, all that. You will not hear a peep of a word out of your game. This is dangerous if you like playing as ghost roles, but if you abhor the thought of it... it's just for you.

The second preference is for admins. You can selectively suppress ghost roles while adminned. This is useful because when you're running an event or doing stuff where you need to offer multiple ghost roles (or you need to focus on a ticket and someone is spamming Xenobiology mob spawns), this is absolutely perfect for suppressing. Same return as the player option, but it checks to see if you are currently adminned via the client.holder variable. This is just because some admins (i'm some admins) don't want to turn in on just in case they forget to turn it off down the line because they actually play the game (lying).

There's probably a much cleaner way to do this code-wise, but I couldn't figure it out. Any help is appreciated. I tested it extensively on my local (even using a guest account), and everything seems to work rather nicely after about an hour of trial-and-error.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Players who want to just alt-tab or maybe chill in deadchat (or have an extreme loathing of ghost roles) can just simply not get any of that. Admins who want to focus on tickets and not have windows pop up to interfere in good administrative work (and be the most annoying thing in the world) can also do that. Everyone is happy.

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
qol: There is now a new preference in Game Preferences, Suppress All Ghost Rolls. If you tick this preference, you will not get a singular window pop-up whenever a Ghost Role is available. Intended for the few who really do need it.
admin: Admins get another additional preference where Suppress All Ghost Roles only works while they are currently in an adminned state. They will still get ghost rolls normally when they are in a deadminned state.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

san7890 committed:
Adds Preferences To Suppress Ghost Role Rolls

Hey there,

Ever since November of 2021, I've wanted something where I could simply not get any ghost roles while adminned. Some people also do not want to get any ghost rolls whatsoever when they play, for it is their personal preference. This PR seeks to resolve both of these issues with two new preferences.

The first preference will show up to everyone, Suppress All Ghost Rolls. It will return on the main proc that pops up the window, does the sound, all that. You will not hear a peep of a word out of your game. This is dangerous if you like playing as ghost roles, but if you abhor the thought of it... it's just for you.

The second preference is for admins. You can selectively suppress ghost roles while adminned. This is useful because when you're running an event or doing stuff where you need to offer multiple ghost roles (or you need to focus on a ticket and someone is spamming Xenobiology mob spawns), this is absolutely perfect for suppressing. Same return as the player option, but it checks to see if you are currently adminned via the client.holder variable.

There's probably a much cleaner way to do this code-wise, but I couldn't figure it out. Any help is appreciated. I tested it extensively on my local (even using a guest account), and everything seems to work rather nicely after about an hour of trial-and-error.

Mothblocks commented on code/modules/client/preferences/ghost.dm:
```diff
@@ -176,3 +176,10 @@
 	category = PREFERENCE_CATEGORY_GAME_PREFERENCES
+
+/// When enabled, prevents any and all ghost role pop-ups.
+/datum/preference/toggle/no_ghost_roles
```
Invert this--make this a ghost roles preference that you turn off.

Mothblocks commented on tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/admin.tsx:
```diff
@@ -29,3 +30,14 @@ export const fast_mc_refresh: FeatureToggle = {
 };
+
+export const no_ghost_roles_as_admin: FeatureToggle = {
+  name: 'Suppress all Ghost Rolls while Adminned',
```
Only capitalize first word

Mothblocks replied:
Typo: Roles

Mothblocks commented on tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/ghost.tsx:
```diff
@@ -130,3 +130,14 @@ export const inquisitive_ghost: FeatureToggle = {
 };
+
+export const no_ghost_roles: FeatureToggle = {
+  name: 'Suppress all Ghost Rolls',
```
Only capitalize first word

Mothblocks replied:
Typo: Roles

Mothblocks commented on code/__HELPERS/game.dm:
```diff
@@ -162,6 +162,14 @@
+	if (candidate_mob.client.prefs.read_preference(/datum/preference/toggle/no_ghost_roles))
+		return
+
+	/// Opt-out for admins whom are currently adminned.
```
Two slashes, this is not a doc comment

san7890 committed:
Review

Standardizes everything to "Ghost Roles"

Indents Stuff

Inverts the check, I think I did what mothblocks wanted?

san7890 committed:
too many exclamation points

boy am i glad i tested this

san7890 committed:
non-necessary paranthesis

yeah don't need those

san7890 committed:
spellchecks descriptions

Mothblocks commented on tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/ghost.tsx:
```diff
@@ -131,13 +131,13 @@ export const inquisitive_ghost: FeatureToggle = {
-  WARNING: If you select this, you will not get any ghost rolls what-so-ever!
-  Every single pop-up WILL be muted for you. This is only useful for those
-  who find ghost roles annoying, use at your own peril.
+    WARNING: If you de-select this, you will not get any ghost rolls pop-ups
```
I don't like the "WARNING:" thing here. I would rather you just explain what it did.

Mothblocks commented on code/modules/client/preferences/admin.dm:
```diff
@@ -52,11 +52,10 @@
 	savefile_identifier = PREFERENCE_PLAYER
-	default_value = FALSE
 
 /datum/preference/toggle/no_ghost_roles_as_admin/is_accessible(datum/preferences/preferences)
```
Didn't change this

san7890 committed:
fixes descriptions and the toggle
